sklearn/linear_model/sag.py

```diff
@@ -0,0 +1,394 @@
+    >>> y = np.random.randn(n_samples)
+    >>> X = np.random.randn(n_samples, n_features)
+    >>> clf = linear_model.Ridge(solver='sag')
+    >>> clf.fit(X, y)
+    ... #doctest: +NORMALIZE_WHITESPACE
+    Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
+          normalize=False, random_state=None, solver='sag', tol=0.001)
+
+    Reference
+    ---------
+    Schmidt, M., Roux, N. L., & Bach, F. (2013).
+    Minimizing finite sums with the stochastic average gradient
+    https://hal.inria.fr/hal-00860051/PDF/sag_journal.pdf
+
+    See also
+    --------
+    Ridge, SGDRegressor, ElasticNet, Lasso, SVR
+    """
+    if max_iter is None:
+        max_iter = 1000
+
+    if check_input:
+        X = check_array(X, dtype=np.float64, accept_sparse='csr', order='C')
+        y = check_array(y, dtype=np.float64, ensure_2d=False, order='C')
+
+    n_samples, n_features = X.shape[0], X.shape[1]
+    alpha = float(alpha) / n_samples
+    fit_intercept = False
+
+    # initialization
+    if sample_weight is None:
+        sample_weight = np.ones(n_samples, dtype=np.float64, order='C')
+
+    # These parameters could be saved between two similar fits, to speed up
+    # convergence (see warm starting behavior of sag_logistic()).
+    # But current Ridge does not have a warm_start parameter, and current
+    # RidgeCV never uses sag solver.
+    coef_ = np.zeros(n_features, dtype=np.float64, order='C')
+    intercept_init = 0.0
+    intercept_sum_gradient_init = 0.0
+    sum_gradient_init = np.zeros(n_features, dtype=np.float64, order='C')
+    gradient_memory_init = np.zeros(n_samples, dtype=np.float64, order='C')
+    seen_init = np.zeros(n_samples, dtype=np.int32, order='C')
+    num_seen_init = 0
+
+    dataset, intercept_decay = make_dataset(X, y, sample_weight, random_state)
+
+    if max_squared_sum is None:
+        max_squared_sum = get_max_squared_sum(X)
+    step_size = get_auto_step_size(max_squared_sum, alpha, "squared",
+                                   fit_intercept)
+
+    if step_size * alpha == 1:
+        raise ZeroDivisionError("Current sag implementation does not handle "
+                                "the case step_size * alpha == 1")
+
+    intercept_, num_seen, n_iter, intercept_sum_gradient = \
+        sag(dataset, coef_.ravel(),
+            intercept_init, n_samples,
+            n_features, tol,
+            max_iter,
+            SquaredLoss(),
+            step_size, alpha,
+            sum_gradient_init.ravel(),
+            gradient_memory_init.ravel(),
+            seen_init.ravel(),
+            num_seen_init,
+            fit_intercept,
+            intercept_sum_gradient_init,
+            intercept_decay,
+            verbose)
+
+    if n_iter == max_iter:
+        warnings.warn("The max_iter was reached which means "
+                      "the coef_ did not converge", ConvergenceWarning)
+
+    return coef_, n_iter
+
+
+def sag_logistic(X, y, sample_weight=None, alpha=1e-4, max_iter=1000,
```

Comment:
> Also, logistic_regression_path calls sag_logistic with alpha=1. / C (logistic.py#L696)

Indeed. I was confused with the parameterization in `SGDClassifier` where I think `alpha = 1 / (C * n_samples)`. To avoid confusion I think we should only use the `C` parameterization in the SAG code. WDYT?